### PR TITLE
BKNDLSS-28749 WebcamPhoto

### PR DESCRIPTION
**Pull request by EXPORTER1700**
create webcam photo component

### CODE REVIEW

File: components/bl-webcamPhoto-component/src/buttons/make-photo.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export function MakePhotoButton({ onClick, text, disabled }) {
+  return (
+    <button
+      onClick={ onClick }
+      className="photo-button"
+      disabled={ disabled }
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/buttons/make-photo.js:6]: disabled={ disabled }>

File: components/bl-webcamPhoto-component/src/index.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  const [modalVisibility, setModalVisibility] = useState(false);
+  const inputRef = useRef();
+
+  const handleClick = useCallback(() => {
+    setModalVisibility(true);
+  }, [modalVisibility]);
+
+  const handleChange = useCallback(async () => {
+    const dataURL = await toBase64(inputRef.current.files[0]);
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/index.js:21]: empty line

File: components/bl-webcamPhoto-component/src/utils/mobile-and-tablet-check.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+export const mobileAndTabletCheck = () => {
+  const toMatch = [
+    /Android/i,
+    /webOS/i,
+    /iPhone/i,
+    /iPad/i,
+    /iPod/i,
+    /BlackBerry/i,
+    /Windows Phone/i,
+  ];
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/utils/mobile-and-tablet-check.js:10]: empty line

File: components/bl-webcamPhoto-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    dataURLToBlob(dataURL)
+      .then(imageBlob => {
+        onSaveImage({ imageBlob });
+      });
+  }, []);
+
+  const isMobile = useMemo(() => {
+    return mobileAndTabletCheck();
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/index.js:29]: checkMobile ?

File: components/bl-webcamPhoto-component/src/utils/to-base64.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const toBase64 = file => new Promise((resolve, reject) => {
+  const reader = new FileReader();
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/utils/to-base64.js:2]: empty line

File: components/bl-webcamPhoto-component/src/index.js (hunk 2 above)
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/index.js:21]: try catch?

File: components/bl-webcamPhoto-component/src/utils/data-url-to-blob.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export async function dataURLToBlob(dataURL) {
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/utils/data-url-to-blob.js:1]: dataURLToBlob
toBase64

into one file 'utils/file.js'

other to 'utils/device.js'

File: components/bl-webcamPhoto-component/src/index.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          reference={ inputRef }
+          disabled={ buttonDisabled }
+        />
+        : <MakePhotoButton
+          onClick={ handleClick }
+          text={ makePhotoButtonLabel }
+          disabled={ buttonDisabled }
+        />
+      }
+      {
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/index.js:47]:
```
      { modalVisibility && (
        <Modal
          setVisibility={ setModalVisibility }
          component={ component }
          eventHandlers={ eventHandlers }
        />
      ) }
```

File: components/bl-webcamPhoto-component/src/modal.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/modal.js:1]: empty line after external modules imports

File: components/bl-webcamPhoto-component/src/modal.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  const videoRef = useRef();
+  const canvasRef = useRef();
+  const [isPhoto, setIsPhoto] = useState(false);
+
+  const handleSnapshot = useCallback(() => {
+    const context = canvasRef.current.getContext('2d');
+    context.drawImage(videoRef.current, 0, 0, 320, 240);
+    setIsPhoto(true);
+  }, [canvasRef]);
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/modal.js:20]: empty line after each handler if it not oneliner

File: components/bl-webcamPhoto-component/src/utils/get-user-media.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  const getStream = stream => {
+    ref.current.srcObject = stream;
+  };
+  const noStream = () => {
+    alert('For the application to work, you must provide access to the camera');
+    close();
+  };
+  navigator.getUserMedia = navigator.getUserMedia ||
+    navigator.webkitGetUserMedia ||
+    navigator.mozGetUserMedia;
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/utils/get-user-media.js:11]: empty line after each logic codeblock

File: components/bl-webcamPhoto-component/src/utils/mobile-and-tablet-check.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+export const mobileAndTabletCheck = () => {
+  const toMatch = [
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/utils/mobile-and-tablet-check.js:2]: regex?

```
 const mobileDeviceRegex  = new Regex('Android|webOS|iPhone...')
 
 mobileDeviceRegex.test(navigator.userAgent)

```

File: components/bl-webcamPhoto-component/styles/index.less [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+      align-items: center;
+
+      video {
+        margin-right: 0;
+        margin-bottom: @bl-customComponent-webcamPhoto-modal-content-video-marginRight;
+      }
+    }
+  }
+}
+
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/styles/index.less:192]: redundant empty lines

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-webcamPhoto-component/src/index.js (hunk 4 above)
Comment (v-excelsior) [reply components/bl-webcamPhoto-component/src/index.js:33]: const isMobile = useMemo(() => {
    return checkMobile();

File: components/bl-webcamPhoto-component/src/modal.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  const handleDone = useCallback(() => {
+    const dataURL = canvasRef.current.toDataURL();
+    dataURLToBlob(dataURL)
+      .then(imageBlob => {
+        onSaveImage({ imageBlob });
+      });
+    handleClose();
+  }, [canvasRef.current]);
+
+  const handleModalClick = useCallback(event => {
```
[the remark this answers is not in the document]
Comment (EXPORTER1700) [reply components/bl-webcamPhoto-component/src/modal.js:35]: Yes, it works correctly
https://drive.google.com/file/d/17lbdZ9x_A5KtYDisr0uFTxdiX90vEDVu/view?usp=sharing

File: components/bl-webcamPhoto-component/src/modal.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        <div className="modal__images">
+          <video
+            ref={ videoRef }
+            width="320"
+            height="240"
+            autoPlay="autoPlay"
+          />
+          <canvas
+            ref={ canvasRef }
+            width="320"
```
[the remark this answers is not in the document]
Comment (EXPORTER1700) [reply components/bl-webcamPhoto-component/src/modal.js:55]: https://ibb.co/ctBGvXW
https://ibb.co/9csLf7x

File: components/bl-webcamPhoto-component/src/utils/device.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
+export function getUserMedia(ref, close) {
+  const getStream = stream => {
+    ref.current.srcObject = stream;
+  };
+  const noStream = () => {
+    alert('For the application to work, you must provide access to the camera');
```
[the remark this answers is not in the document]
Comment (EXPORTER1700) [reply components/bl-webcamPhoto-component/src/utils/device.js]: https://drive.google.com/file/d/1k8VnaJ8pdyYfdJRwIekNOlM0qnT8PoM7/view?usp=sharing
Вот видео работы.

File: components/bl-webcamPhoto-component/src/popup/options.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+export const PopupOptionsMap = {
+  noYetPermission: {
+    title      : 'Give access to the camera',
+    description: 'To work properly, you must allow the app to access the camera.'
+  },
+  noPermission   : {
+    title      : 'No camera access',
+    description: 'Blocked app access to the camera.Follow the link for detailed information on how to grant access.',
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/popup/options.js:8]: space after `camera.`

File: components/bl-webcamPhoto-component/src/popup/options.js [hunk 17]
```diff
@@ -0,0 +1,11 @@
+export const PopupOptionsMap = {
```
Comment (v-excelsior) [inline components/bl-webcamPhoto-component/src/popup/options.js:1]: This and popup/index.js could be combined into one file moved to src/popup.js

Also, add please screenshots for popups

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-webcamPhoto-component/src/buttons/done.js [hunk 18]
```diff
@@ -0,0 +1,10 @@
+export function DoneButton({ onClick, text, disabled }) {
+  return (
+    <button
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/buttons/done.js:3]: to single line

File: components/bl-webcamPhoto-component/src/buttons/make-photo.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+export function MakePhotoButton({ onClick, text, disabled }) {
+  return (
+    <button
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/buttons/make-photo.js:3]: to single line, please apply this for all small components

File: components/bl-webcamPhoto-component/src/helpers/device.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  const noStream = () => {
+    setPopupOptions(PopupOptionsMap.noPermission);
+  };
+
+  navigator.getUserMedia = navigator.getUserMedia ||
+    navigator.webkitGetUserMedia ||
+    navigator.mozGetUserMedia;
+
+  if (navigator.getUserMedia) {
+    navigator.webkitGetUserMedia(
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/helpers/device.js:18]: `navigator.getUserMedia(...)`?

File: components/bl-webcamPhoto-component/src/helpers/file.js [hunk 21]
```diff
@@ -0,0 +1,12 @@
+export async function dataURLToBlob(dataURL) {
+  const response = await fetch(dataURL);
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/helpers/file.js:2]: wouldn't it better to use sync code?

https://stackoverflow.com/questions/23150333/html5-javascript-dataurl-to-blob-blob-to-dataurl/30407959#30407959

File: components/bl-webcamPhoto-component/src/index.js [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+  const handleClick = useCallback(() => {
+    setModalVisibility(true);
+  }, [modalVisibility]);
+
+  const handleChange = useCallback(async () => {
+    try {
+      const dataURL = await toBase64(inputRef.current.files[0]);
+
+      dataURLToBlob(dataURL)
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/index.js:23]:
```
 const imageBlob = await dataURLToBlob(dataURL)

onSaveImage({ imageBlob });
```

File: components/bl-webcamPhoto-component/src/index.js [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+        .then(imageBlob => {
+          onSaveImage({ imageBlob });
+        });
+    } catch (error) {
+      console.error(`Webcam Photo: ${ error.message }`);
+    }
+  }, []);
+
+  const isMobile = useMemo(() => {
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/index.js:32]: we do not need to calculate the value for each component, the value won't be changed

File: components/bl-webcamPhoto-component/src/index.js [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    return checkMobile();
+  }, []);
+
+  return (
+    <div className="bl-customComponent-webcamPhoto">
+      { isMobile
+        ? <UploadButton
+          onChange={ handleChange }
+          text={ uploadButtonLabel }
+          reference={ inputRef }
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/index.js:42]: `inputRef={ inputRef }` could be more readable

File: components/bl-webcamPhoto-component/src/index.js [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+import { dataURLToBlob, toBase64 } from './helpers/file';
+import { checkMobile } from './helpers/device';
+import { Modal } from './modal';
+
+export default function WebcamPhoto({ component, eventHandlers }) {
+  const { uploadButtonLabel, makePhotoButtonLabel, buttonDisabled } = component;
+  const { onSaveImage } = eventHandlers;
+
+  const [modalVisibility, setModalVisibility] = useState(false);
+  const inputRef = useRef();
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/index.js:13]: `const fileInputRef = useRef();`

File: components/bl-webcamPhoto-component/src/popup/index.js [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+
+export function Popup({ options }) {
+  const { title, description, link } = options;
+
+  const handleClick = useCallback(event => {
+    event.stopPropagation();
+  }, [event]);
+
+  return (
+    <div className="popup" onClick={ handleClick }>
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/popup/index.js:11]:
```
 <div className="popup" onClick={ handleClick }>
      <p className="popup__title">{ title }</p>
      <p className="popup__description">{ description }</p>
     
      { link && (
        <a className="popup__link" href={ link } target="_blank">nfo...</a>
      ) }
    </div>
```

File: components/bl-webcamPhoto-component/src/popup/index.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+      <p className="popup__title">{ title }</p>
+      <p className="popup__description">{ description }</p>
+      { link && (
+        <a
+          className="popup__link"
+          href={ link }
+          target="_blank">
+          Info...
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/popup/index.js:19]: shouldn't the `Info...` label come from the settings?

File: components/bl-webcamPhoto-component/src/popup/options.js [hunk 28]
```diff
@@ -0,0 +1,11 @@
...
+  noYetPermission: {
+    title      : 'Give access to the camera',
```
Comment (Valodya) [inline components/bl-webcamPhoto-component/src/popup/options.js:3]: is there any way to change these messages for a particular component instance?

File: components/bl-webcamPhoto-component/src/index.js (hunk 23 above)
Comment (EXPORTER1700) [reply components/bl-webcamPhoto-component/src/index.js]: moved isMobile outside of the component, right?

File: components/bl-webcamPhoto-component/src/popup/options.js (hunk 17 above)
Comment (EXPORTER1700) [reply components/bl-webcamPhoto-component/src/popup/options.js]: image

**Pull request merged**